Moved EventCategory definition to review-protocol

syncpark opened the pull request:
Close #304 
- Moved `EventCategory` definition to review-protocol crate
- `HttpThreat` category is changed to **obsoleted** and replaced existing `HttpThreat` category value to `Reconnaissance`

msk commented on src/event/conn.rs:
```diff
@@ -7,8 +7,11 @@ use std::{
+use super::{common::Match, TriagePolicy, TriageScore, MEDIUM};
+use crate::{
+    event::common::{triage_scores_to_string, vector_to_string},
+    types::EventCategory,
```
The `review-database` crate still retains its own `EventCategory` enum. I'm not clear on why all usages of it are being replaced with `review_protocol::EventCategory`. Directly using `review_protocol::EventCategory` throughout `review-database` could potentially create unnecessary coupling between the two crates.

Could you explain the reasoning behind using `review_protocol::EventCategory` instead of `review_database::EventCategory` within the review-database codebase? What are the advantages you see in this approach?

msk commented:
@minshao Could you check if changing the definition of `enum EventCategory` requires any database migration? Other than that, I’m good with this PR.

minshao commented:
> @minshao Could you check if changing the definition of `enum EventCategory` requires any database migration? Other than that, I’m good with this PR.

Since  `EventCategory::HttpThreat` is removed, existing database entries contains `EventCategory::HttpThreat` needs to be updated. That is, all entries in `TriagePolicy` map needs to be migrated.

minshao commented:
I'd like to clarify the intention behind the recent changes to the `HttpThreat` field. 

As a reminder, database migration is typically only necessary when there are changes to the underlying database fields. In the case of deprecation, where a field is no longer used or is being replaced, migration is usually not required.

However, in the latest commit (commit ID: 4332d9989c10c45a00440d2166175dc4c4c91c42):
* the CHANGELOG states that `HttpThreat` **is removed**. 
* Upon reviewing the code, I noticed that `HttpThreat` is only marked as **"obsoleted"** and **not** actually removed. 
* Furthermore, the migration code is updating existing entries to **replace** `HttpThreat` with `Reconnaissance`.

I'm having trouble understanding the intended outcome here. @syncpark Should we be removing HttpThreat entirely, or is it intended to remain as an obsolete field? I'd appreciate any clarification on this matter to ensure a smooth transition.

syncpark commented:
> I'd like to clarify the intention behind the recent changes to the `HttpThreat` field.
> 
> As a reminder, database migration is typically only necessary when there are changes to the underlying database fields. In the case of deprecation, where a field is no longer used or is being replaced, migration is usually not required.
> 
> However, in the latest commit (commit ID: [4332d99](https://github.com/petabi/review-database/commit/4332d9989c10c45a00440d2166175dc4c4c91c42)):
> 
> * the CHANGELOG states that `HttpThreat` **is removed**.
> * Upon reviewing the code, I noticed that `HttpThreat` is only marked as **"obsoleted"** and **not** actually removed.
> * Furthermore, the migration code is updating existing entries to **replace** `HttpThreat` with `Reconnaissance`.
> 
> I'm having trouble understanding the intended outcome here. @syncpark Should we be removing HttpThreat entirely, or is it intended to remain as an obsolete field? I'd appreciate any clarification on this matter to ensure a smooth transition.

I'm sorry for causing the confusion.

The initial intention was to delete it. However, it seemed like it would be a problem to delete the enum value while the programs that use that value had not yet been modified, so I modified it with **obsolete** while writing the migration code. Deleting this value will be done after all program modifications have been completed.

The `HttpThreat` category will be removed in the near future, all the existing `HttpThreat` category values are replaced to `Reconnaissance`. And a new field `category` that is inserted in all the detected event structures will be used instead of  hardcoded category value.

I modified the CHANGELOG.

minshao commented:
> I'm sorry for causing the confusion.
> 
> The initial intention was to delete it. However, it seemed like it would be a problem to delete the enum value while the programs that use that value had not yet been modified, so I modified it with **obsolete** while writing the migration code. Deleting this value will be done after all program modifications have been completed.
> 
> The `HttpThreat` category will be removed in the near future, all the existing `HttpThreat` category values are replaced to `Reconnaissance`. And a new field `category` that is inserted in all the detected event structures will be used instead of hardcoded category value.
> 
> I modified the CHANGELOG.

No worries, thanks for the clarification, that's very helpful. I'll proceed accordingly.

minshao commented on src/types.rs:
```diff
@@ -51,7 +52,7 @@ pub enum EventCategory {
     Exfiltration,
     Impact,
-    HttpThreat,
+    HttpThreat, // Obsoleted. Do not use this category
```
Instead of comment, [deprecated attribute](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-deprecated-attribute) could be more helpful.
```suggestion
    #[deprecated(note = "Users should use `EventCategory::Reconnaissance` instead")]
    HttpThreat, 
```

minshao replied:
After addressing the clippy warnings, it looks good to me.

syncpark replied:
With the directive `#[deprecated(note = "...")]`, `HttpThreat` enum member always makes clippy error like following:
```rust
error: use of deprecated unit variant `types::EventCategory::HttpThreat`: Users should use `EventCategory::Reconnaissance` instead
  --> src/types.rs:56:5
   |
56 |     HttpThreat, // Obsoleted. Do not use this category
   |     ^^^^^^^^^^
   |
   = note: `-D deprecated` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(deprecated)]`
```

The directive to indicate that it is `deprecated` creates a deprecated clippy error. I haven't found a solution yet.

I changed it to comment now.

minshao replied:
Sorry, I didn't realize that proc macro would still cause the warning even with `#[allow(deprecated)]`. The following should do the trick.
```diff
diff --git a/src/types.rs b/src/types.rs
index 0a7cb57..0b41480 100644
--- a/src/types.rs
diff --git a/src/types.rs b/src/types.rs
index 0a7cb57..0b41480 100644
--- a/src/types.rs
+++ b/src/types.rs
@@ -1,3 +1,4 @@
+#![allow(deprecated)]
 use std::{cmp::Ordering, net::IpAddr, ops::RangeInclusive};
 
 use anyhow::Result;
@@ -52,7 +53,8 @@ pub enum EventCategory {
     CommandAndControl,
     Exfiltration,
     Impact,
-    HttpThreat, // Obsoleted. Do not use this category
+    #[deprecated(note = "Users should use `EventCategory::Reconnaissance` instead")]
+    HttpThreat,
 }
 
 #[derive(Deserialize)]
```

minshao commented on CHANGELOG.md:
```diff
@@ -33,6 +33,10 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 - Used `FromPrimitive` and `ToPrimitive` for converting `EventCategory` instead of
   manually implementing `TryFrom`.
+- `EventCategory` definition is moved to `review-protocol`.
+  `HttpThreat` category is obsoleted and replaced to `Reconnaissance`.
```
* Since `HttpThreat` is only deprecated at this stage, it should be mentioned in deprecated section in CHANGELOG, see line 315 of CHANGELOG.
* "replaced to" -> "replaced with"?

minshao commented on src/migration.rs:
Since `HttpThreat` is not removed yet and is still part of the Public API, it is not necessary to migrate all existing data entries at this point.
